src/types/string.rs

```diff
@@ -144,6 +144,33 @@ impl PyString {
         unsafe { py.from_owned_ptr(ffi::PyUnicode_FromStringAndSize(ptr, len)) }
     }
 
+    /// Intern the given string
+    ///
+    /// This will return a reference to the same Python string object if called repeatedly with the same string.
+    ///
+    /// Note that a temporary allocation is required if the given string is not already null-terminated.
+    ///
+    /// Panics if out of memory or null byte within string.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use pyo3::{types::PyString, Python};
+    ///
+    /// # Python::with_gil(|py| {
+    /// let py_str = PyString::intern(py, "foobar\0");
+    /// assert_eq!(py_str.to_str().unwrap(), "foobar");
+    /// # });
+    /// ```
+    pub fn intern<'p>(py: Python<'p>, s: &str) -> &'p PyString {
+        if let Ok(s) = CStr::from_bytes_with_nul(s.as_bytes()) {
+            unsafe { py.from_owned_ptr(ffi::PyUnicode_InternFromString(s.as_ptr())) }
+        } else {
+            let s = CString::new(s).unwrap();
+            unsafe { py.from_owned_ptr(ffi::PyUnicode_InternFromString(s.as_ptr())) }
+        }
+    }
```

Comment:
Note that the C function creates a PyString and interns that:

```c
PyObject *
PyUnicode_InternFromString(const char *cp)
{
    PyObject *s = PyUnicode_FromString(cp);
    if (s == NULL)
        return NULL;
    PyUnicode_InternInPlace(&s);
    return s;
}
```

Perhaps it would be better to do what that function does:

```rust
pub fn intern<'p>(py: Python<'p>, s: &str) -> &'p PyString {
        let pystring = PyUnicode_FromStringAndSize(s.as_ptr() as *const _, s.len());
        PyUnicode_InternInPlace(pystring);
        py.from_owned_ptr(pystring)
}
```

Reply:
Nice, that definitely makes more sense.

Reply:
This is what I started with, but as written in https://github.com/PyO3/pyo3/pull/2268#issuecomment-1086797378, I consider the current implementation in CPython to be a performance bug as the `PyUnicode_InternFromString` does not need to literally call `PyUnicode_FromString` followed by `PyUnicode_InternInPlace`.

It could rather look up a possibly already interned string using only the null-terminated string and just increment its reference count without any allocation at all. The way it is written ATM would pick this up when `PyUnicode_InternFromString` is fixed.